pkg/services/serviceaccounts/manager/service_test.go

```diff
@@ -4,17 +4,146 @@ import (
+}
+
+// DeleteServiceAccount is a fake deleting a service account.
+func (f *FakeServiceAccountStore) DeleteServiceAccount(ctx context.Context, orgID, serviceAccountID int64) error {
+	return f.ExpectedError
+}
+
+// GetAPIKeysMigrationStatus is a fake getting the api keys migration status.
+func (f *FakeServiceAccountStore) GetAPIKeysMigrationStatus(ctx context.Context, orgID int64) (*serviceaccounts.APIKeysMigrationStatus, error) {
+	return f.ExpectedServiceAccountMigrationStatus, f.ExpectedError
+}
+
+// HideApiKeysTab is a fake hiding the api keys tab.
+func (f *FakeServiceAccountStore) HideApiKeysTab(ctx context.Context, orgID int64) error {
+	return f.ExpectedError
+}
+
+// MigrateApiKeysToServiceAccounts is a fake migrating api keys to service accounts.
+func (f *FakeServiceAccountStore) MigrateApiKeysToServiceAccounts(ctx context.Context, orgID int64) error {
+	return f.ExpectedError
+}
+
+// MigrateApiKey is a fake migrating an api key to a service account.
+func (f *FakeServiceAccountStore) MigrateApiKey(ctx context.Context, orgID int64, keyId int64) error {
+	return f.ExpectedError
+}
+
+// RevertApiKey is a fake reverting an api key to a service account.
+func (f *FakeServiceAccountStore) RevertApiKey(ctx context.Context, saId int64, keyId int64) error {
+	return f.ExpectedError
+}
+
+// ListTokens is a fake listing tokens.
+func (f *FakeServiceAccountStore) ListTokens(ctx context.Context, query *serviceaccounts.GetSATokensQuery) ([]apikey.APIKey, error) {
+	return f.ExpectedApiKeys, f.ExpectedError
+}
+
+// RevokeServiceAccountToken is a fake revoking a service account token.
+func (f *FakeServiceAccountStore) RevokeServiceAccountToken(ctx context.Context, orgId, serviceAccountId, tokenId int64) error {
+	return f.ExpectedError
+}
+
+// AddServiceAccountToken is a fake adding a service account token.
+func (f *FakeServiceAccountStore) AddServiceAccountToken(ctx context.Context, serviceAccountID int64, cmd *serviceaccounts.AddServiceAccountTokenCommand) error {
+	return f.ExpectedError
+}
+
+// DeleteServiceAccountToken is a fake deleting a service account token.
+func (f *FakeServiceAccountStore) DeleteServiceAccountToken(ctx context.Context, orgID, serviceAccountID, tokenID int64) error {
+	return f.ExpectedError
+}
+
+// GetUsageMetrics is a fake getting usage metrics.
+func (f *FakeServiceAccountStore) GetUsageMetrics(ctx context.Context) (*serviceaccounts.Stats, error) {
+	return f.ExpectedStats, f.ExpectedError
+}
+
+type SecretsCheckerFake struct {
+	ExpectedError error
+}
+
+func (f *SecretsCheckerFake) CheckTokens(ctx context.Context) error {
+	return f.ExpectedError
+}
+
 func TestProvideServiceAccount_DeleteServiceAccount(t *testing.T) {
-	t.Run("should call store function", func(t *testing.T) {
-		storeMock := &tests.ServiceAccountsStoreMock{Calls: tests.Calls{}}
-		svc := ServiceAccountsService{store: storeMock}
-		err := svc.DeleteServiceAccount(context.Background(), 1, 1)
+	storeMock := newServiceAccountStoreFake()
```

Comment:
nit: this test seems a little bit odd to me, and I think I don't quite understand what we're aiming to test here.

Firstly, it only tests the mock, which imo is not needed. Secondly, it tests both adding and deletion in one test function, but none of the other SA functionality.

Reply:
talking to the backend team and kalle we should test the struct of interest and mock away the rest to have unit tests only testing the functionality of the struct implementation.
In this case it is the `ServiceAccountsService{}` that is the real thing we want to test and mock away the rest.

Your second point of it only testing deletion is correct. This did not test for anything else than the deletion as well before this refactoring. I could certainly add testing functionality to this refactoring work. If possible we could add another issue for it instead, this would be preferred from my place as we really want to move the store functionality to private as fast as possible. Both in terms of merge conflicts with this PR and both to not have it exposed.

Reply:
adding another issue to refactoring backlog for service accounts to make tests for this service.
https://github.com/grafana/grafana/issues/60240